pkg/kubectl/cmd/apply_test.go

```diff
@@ -531,3 +534,89 @@ func TestApplyNULLPreservation(t *testing.T) {
 		t.Fatal("No server-side patch call detected")
 	}
 }
+
+func readTPREntryFromFile(t *testing.T, filename string) []byte {
+	raw := readBytesFromFile(t, filename)
+	obj := &unstructured.Unstructured{}
+	if err := runtime.DecodeInto(testapi.Extensions.Codec(), raw, obj); err != nil {
+		t.Fatal(err)
+	}
+
+	objJSON, err := runtime.Encode(testapi.Extensions.Codec(), obj)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return objJSON
+}
+
+// TestApplyIdempotentTPREntry check repeated apply operations on TPR entries
+// without any change.
+func TestApplyIdempotentTPREntry(t *testing.T) {
+	initTestErrorHandler(t)
+	name := "test-tpr-entry"
+	path := "/namespaces/test/testtprs/" + name
+	curr := readTPREntryFromFile(t, filenameIdempotentTPREntry)
+
+	verifiedPatch := false
+
+	f, tf, _, _ := cmdtesting.NewAPIFactory()
+	tf.Printer = &testPrinter{}
+	tf.UnstructuredClient = &fake.RESTClient{
+		APIRegistry:          api.Registry,
+		NegotiatedSerializer: unstructuredSerializer,
+		Client: fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
+			switch p, m := req.URL.Path, req.Method; {
+			case p == path && m == "GET":
+				body := ioutil.NopCloser(bytes.NewReader(curr))
+				return &http.Response{
+					StatusCode: 200,
+					Header:     defaultHeader(),
+					Body:       body}, nil
+			case p == path && m == "PATCH":
+				patch, err := ioutil.ReadAll(req.Body)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				patchMap := map[string]interface{}{}
+				if err := json.Unmarshal(patch, &patchMap); err != nil {
+					t.Fatal(err)
+				}
+
+				annotationsMap := walkMapPath(t, patchMap, []string{"metadata", "annotations"})
+				if _, ok := annotationsMap[annotations.LastAppliedConfigAnnotation]; !ok {
+					t.Fatalf("patch does not contain annotation:\n%s\n", patch)
+				}
+
+				verifiedPatch = true
+
+				body := ioutil.NopCloser(bytes.NewReader(curr))
+				return &http.Response{
+					StatusCode: 200,
+					Header:     defaultHeader(),
+					Body:       body}, nil
+			default:
+				t.Fatalf("unexpected request: %#v\n%#v", req.URL, req)
+				return nil, nil
+			}
+		}),
+	}
+
+	tf.Namespace = "test"
+	buf := bytes.NewBuffer([]byte{})
+	errBuf := bytes.NewBuffer([]byte{})
+
+	cmd := NewCmdApply(f, buf, errBuf)
+	cmd.Flags().Set("filename", filenameIdempotentTPREntry)
```

Comment:
you should not give the apply command a fixture that already has a last-applied annotation in it. that should only exist server-side.

there are two testcases here:

1. create case:
  * apply -f file-without-annotation.yaml
  * server responds with 404
  * create API call is made
  * verify submitted content matches the fixture content *with an additional last-applied annotation added*

2. idempotent apply:
  * apply -f file-without-annotation.yaml
  * server responds with 200 response with object matching the fixture *with a last-applied annotation added*
  * apply no-ops (I forget if it sends an empty patch or skips patching in this case)
  * verify patch content is empty and content-type is generic json merge patch